client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java

```diff
@@ -247,10 +282,55 @@ private void requestExecutorMemory(long leastMem) {
     }
   }
 
+  public List<AddBlockEvent> buildBlockEvents(List<ShuffleBlockInfo> shuffleBlockInfoList) {
+    long totalSize = 0;
+    long memoryUsed = 0;
+    List<AddBlockEvent> events = new ArrayList<>();
+    List<ShuffleBlockInfo> shuffleBlockInfosPerEvent = Lists.newArrayList();
+    for (ShuffleBlockInfo sbi : shuffleBlockInfoList) {
+      totalSize += sbi.getSize();
+      memoryUsed += sbi.getFreeMemory();
+      shuffleBlockInfosPerEvent.add(sbi);
+      // split shuffle data according to the size
+      if (totalSize > sendSizeLimit) {
+        LOG.info("Build event with " + shuffleBlockInfosPerEvent.size()
+            + " blocks and " + totalSize + " bytes");
+        final long _memoryUsed = memoryUsed;
+        events.add(
+            new AddBlockEvent(taskId, shuffleBlockInfosPerEvent, () -> freeAllocatedMemory(_memoryUsed))
+        );
+        shuffleBlockInfosPerEvent = com.google.common.collect.Lists.newArrayList();
+        totalSize = 0;
+        memoryUsed = 0;
+      }
+    }
+    if (!shuffleBlockInfosPerEvent.isEmpty()) {
+      LOG.info("Build event with " + shuffleBlockInfosPerEvent.size()
+          + " blocks and " + totalSize + " bytes");
+      final long _memoryUsed = memoryUsed;
+      events.add(
+          new AddBlockEvent(taskId, shuffleBlockInfosPerEvent, () -> freeAllocatedMemory(_memoryUsed))
+      );
+    }
+    return events;
+  }
+
   @Override
   public long spill(long size, MemoryConsumer trigger) {
-    // there is no spill for such situation
-    return 0;
+    List<AddBlockEvent> events = buildBlockEvents(clear());
+    AtomicLong estimatedMemorySize = new AtomicLong(0L);
+    events.stream().forEach(x -> {
+      spillAsyncFunc.apply(x);
+      estimatedMemorySize.addAndGet(x.getShuffleDataInfoList()
+          .stream()
+          .map(k -> k.getFreeMemory())
+          .reduce((a, b) -> a + b).get()
+      );
+    });
+    taskMemoryManager.showMemoryUsage();
+    LOG.info("Prepare sending data to remote shuffle server, estimated free memory size: {}",
+        estimatedMemorySize.get());
+    return 0L;
```

Comment:
> I believe it's reasonable to wait some time to get some data sent, and releases memory to executor?

Actually, the async or sync wait is all OK for this part. 

If using sync wait, when `acquireExecutionMemory` is invoked and memory is not enough, the taskMemoryManager will re-request memory after invoking spill func from several `memoryConsumers` if the released size is not zero.

But if using async wait, the re-request memory logic will be raised by the uniffle's `WriterBufferManager#requestExecutorMemory`.

But it's better to support async wait in `spill`

Reply:
> and saw that it didn't re-request more memory when spill releases zero. So I'm a bit worried whether it will work correctly if zero is returned.

If 0, it won't. But the re-request logic will be handled by uniffle's `WriterBufferManager#requestExecutorMemory.`

Reply:
I have updated the code to support sync wait with timeout.

Reply:
+1 for @advancedxy https://github.com/apache/incubator-uniffle/pull/714#discussion_r1135567277

Reply:
I have updated, PTAL again @smallzhongfeng Thanks